src/main/java/com/aws/greengrass/clientdevices/auth/iot/registry/ThingRegistry.java

```diff
@@ -5,33 +5,101 @@
 
 package com.aws.greengrass.clientdevices.auth.iot.registry;
 
+import com.aws.greengrass.clientdevices.auth.api.DomainEvents;
 import com.aws.greengrass.clientdevices.auth.exception.CloudServiceInteractionException;
 import com.aws.greengrass.clientdevices.auth.iot.Certificate;
 import com.aws.greengrass.clientdevices.auth.iot.IotAuthClient;
 import com.aws.greengrass.clientdevices.auth.iot.Thing;
+import com.aws.greengrass.clientdevices.auth.iot.events.ThingEvent;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.LinkedHashMap;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import javax.inject.Inject;
 
 public class ThingRegistry {
-    // holds mapping of thingName to IoT Certificate ID;
+    // holds mapping of thingName to IoT Certificate IDs;
     // size-bound by default cache size, evicts oldest written entry if the max size is reached
-    private final Map<String, String> registry = Collections.synchronizedMap(
-            new LinkedHashMap<String, String>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
+    private final Map<String, List<String>> registry = Collections.synchronizedMap(
+            new LinkedHashMap<String, List<String>>(RegistryConfig.REGISTRY_CACHE_SIZE, 0.75f, false) {
                 @Override
                 protected boolean removeEldestEntry(Map.Entry eldest) {
                     return size() > RegistryConfig.REGISTRY_CACHE_SIZE;
                 }
             });
 
     private final IotAuthClient iotAuthClient;
+    private final DomainEvents domainEvents;
 
     @Inject
-    public ThingRegistry(IotAuthClient iotAuthClient) {
+    public ThingRegistry(IotAuthClient iotAuthClient, DomainEvents domainEvents) {
         this.iotAuthClient = iotAuthClient;
+        this.domainEvents = domainEvents;
+    }
+
+    /**
+     * Retrieve a Thing based on ThingName.
+     *
+     * @param thingName ThingName
+     * @return Thing domain object
+     */
+    public Thing getThing(String thingName) {
+        List<String> certificateIds = registry.get(thingName);
+        if (certificateIds != null) {
+            return new Thing(thingName, certificateIds);
+        }
+        return null;
+    }
+
+    /**
+     * Attach a certificate to Thing.
+     * </p>
+     * If the provided Thing does not exist, it will be created.
+     * @param thingName     ThingName.
+     * @param certificateId Certificate ID to attach.
+     */
+    public synchronized void attachCertificateToThing(String thingName, String certificateId) {
+        List<String> certificateIds = registry.get(thingName);
+
+        if (certificateIds != null && certificateIds.contains(certificateId)) {
```

Comment:
If we were using this value object then this messages could be asked to the `Thing` 

```java
thing.hasCertificateId(certificateId)
```

Also passing the value object directly would guarantee we don't have to worry about creating a new `ArrayList` since we would already have a thing. This method would become more of an  `update`

Reply:
Let's chat about the best way to handle this one.

I am nervous about passing around Thing objects, because I am afraid of how they will be used. I want to treat anything provided by the registry as an immutable snapshot. They should be one time use objects, which are thrown away.

The reason for that is, that I don't want the registry to need to maintain references to every single Thing object that is ever returned. The Registry has no way of knowing if an external caller still has a reference to an object, so it doesn't know if it can "throw it away".

It would perhaps be better if only the Registry can instantiate Thing objects. And if we go that route, then we can't base an update API using it.